packager/packager.cc

```diff
@@ -0,0 +1,665 @@
+                // Create a module for the imported package that sets up constant references to exported items.
+                // Use proper loc information on the module name so that `import Foo` displays in the results of LSP
+                // Find All References on `Foo`.
+                importedPackages.emplace_back(ast::MK::Module(imported.loc, core::Loc(ctx.file, imported.loc),
+                                                              imported.getFullNameLiteral(imported.loc), {},
+                                                              std::move(exportedItemsCopy)));
+            }
+        }
+    }
+
+    // Include Kernel in the package.
+    importedPackages.emplace_back(ast::MK::Send1(core::LocOffsets::none(), ast::MK::Self(core::LocOffsets::none()),
+                                                 core::Names::include(), name2Expr(core::Names::Constants::Kernel())));
+
+    auto packageNamespace =
+        ast::MK::Module(core::LocOffsets::none(), core::Loc(ctx.file, core::LocOffsets::none()),
+                        name2Expr(package.name.mangledName, name2Expr(core::Names::Constants::PackageRegistry())), {},
+                        std::move(importedPackages));
+
+    auto rootKlass = ast::cast_tree<ast::ClassDef>(file.tree);
+    ENFORCE(rootKlass != nullptr);
+    rootKlass->rhs.emplace_back(move(packageNamespace));
+    return file;
+}
+
+ast::ParsedFile rewritePackagedFile(core::Context ctx, ast::ParsedFile file, core::NameRef packageMangledName) {
+    auto rootKlass = ast::cast_tree<ast::ClassDef>(file.tree);
+    ENFORCE(rootKlass != nullptr);
+    auto moduleWrapper =
+        ast::MK::Module(core::LocOffsets::none(), core::Loc(ctx.file, core::LocOffsets::none()),
+                        name2Expr(packageMangledName, name2Expr(core::Names::Constants::PackageRegistry())), {},
+                        std::move(rootKlass->rhs));
+    rootKlass->rhs.clear();
+    rootKlass->rhs.emplace_back(move(moduleWrapper));
+    return file;
+}
+
+// We can't run packages without having all package ASTs. Assert that they are all present.
+bool checkContainsAllPackages(const core::GlobalState &gs, const vector<ast::ParsedFile> &files) {
+    UnorderedSet<core::FileRef> allPackages;
+    for (u4 i = 1; i < gs.filesUsed(); i++) {
+        core::FileRef fref(i);
+        if (fref.data(gs).sourceType == core::File::Type::Package) {
+            allPackages.insert(fref);
+        }
+    }
+
+    for (const auto &f : files) {
+        if (f.file.data(gs).sourceType == core::File::Type::Package) {
+            allPackages.erase(f.file);
+        }
+    }
+    return allPackages.empty();
+}
+
+} // namespace
+
+vector<ast::ParsedFile> Packager::run(core::GlobalState &gs, WorkerPool &workers, vector<ast::ParsedFile> files) {
+    Timer timeit(gs.tracer(), "packager");
+    // Ensure files are in canonical order.
+    fast_sort(files, [](const auto &a, const auto &b) -> bool { return a.file < b.file; });
```

Comment:
Why is this important? This could be slow so I just want to understand why.

Reply:
Deterministic errors. That's it.